src/main/scala/org/broadinstitute/hail/expr/FunctionRegistry.scala

```diff
@@ -362,12 +602,481 @@ object FunctionRegistry {
+          null
+        else
+          f(xi.asInstanceOf[T], yi.asInstanceOf[T])
+      }
+    })(arrayHr(hrboxedt), arrayHr(hrboxedt), arrayHr(hrboxeds))
+  }
+
+  registerMethod("toInt", (s: String) => s.toInt)
+  registerMethod("toLong", (s: String) => s.toLong)
+  registerMethod("toFloat", (s: String) => s.toFloat)
+  registerMethod("toDouble", (s: String) => s.toDouble)
+
+  registerMethod("toInt", (b: Boolean) => b.toInt)
+  registerMethod("toLong", (b: Boolean) => b.toLong)
+  registerMethod("toFloat", (b: Boolean) => b.toFloat)
+  registerMethod("toDouble", (b: Boolean) => b.toDouble)
+
+  def registerNumericType[T]()(implicit ev: Numeric[T], hrt: HailRep[T]) {
+    registerNumeric("+", ev.plus)
+    registerNumeric("-", ev.minus)
+    registerNumeric("*", ev.times)
+    registerNumeric("/", (x: T, y: T) => ev.toDouble(x) / ev.toDouble(y))
+
+    registerMethod("abs", ev.abs _)
+    registerMethod("signum", ev.signum _)
+
+    register("-", ev.negate _)
+    register("fromInt", ev.fromInt _)
+
+    registerMethod("toInt", ev.toInt _)
+    registerMethod("toLong", ev.toLong _)
+    registerMethod("toFloat", ev.toFloat _)
+    registerMethod("toDouble", ev.toDouble _)
+  }
+
+  registerNumericType[Int]()
+  registerNumericType[Long]()
+  registerNumericType[Float]()
+  registerNumericType[Double]()
+
+  register("==", (a: Any, b: Any) => a == b)(TTHr, TUHr, boolHr)
+  register("!=", (a: Any, b: Any) => a != b)(TTHr, TUHr, boolHr)
+
+  def registerOrderedType[T]()(implicit ord: Ordering[T], hrt: HailRep[T]) {
+    val hrboxedt = new HailRep[Any] {
+      def typ: Type = hrt.typ
+    }
+
+    register("<", ord.lt _)
+    register("<=", ord.lteq _)
+    register(">", ord.gt _)
+    register(">=", ord.gteq _)
+
+    registerMethod("min", ord.min _)
+    registerMethod("max", ord.max _)
+
+    registerMethod("sort", (a: IndexedSeq[Any]) => a.sorted(extendOrderingToNull(ord)))(arrayHr(hrboxedt), arrayHr(hrboxedt))
+    registerMethod("sort", (a: IndexedSeq[Any], ascending: Boolean) =>
+      a.sorted(extendOrderingToNull(
+        if (ascending)
+          ord
+        else
+          ord.reverse))
+    )(arrayHr(hrboxedt), boolHr, arrayHr(hrboxedt))
+
+    registerLambdaMethod("sortBy", (a: IndexedSeq[Any], f: (Any) => Any) =>
+      a.sortBy(f)(extendOrderingToNull(ord))
+    )(arrayHr(TTHr), unaryHr(TTHr, hrboxedt), arrayHr(TTHr))
+
+    registerLambdaMethod("sortBy", (a: IndexedSeq[Any], f: (Any) => Any, ascending: Boolean) =>
+      a.sortBy(f)(extendOrderingToNull(
+        if (ascending)
+          ord
+        else
+          ord.reverse))
+    )(arrayHr(TTHr), unaryHr(TTHr, hrboxedt), boolHr, arrayHr(TTHr))
+  }
+
+  registerOrderedType[Int]()
+  registerOrderedType[Long]()
+  registerOrderedType[Float]()
+  registerOrderedType[Double]()
+  registerOrderedType[String]()
+
+  register("%", (x: Int, y: Int) => x % y)
+  register("%", (x: Long, y: Long) => x % y)
+  register("+", (x: String, y: Any) => x + y)(stringHr, TTHr, stringHr)
```

Comment:
I can't support that because then String + String is ambiguous.  I updated the docs on this point.